jcl/src/java.base/share/classes/com/ibm/oti/vm/VM.java

```diff
@@ -375,25 +374,42 @@ private static void shutdown() {
  * @return		a String array containing the optionString part of command line arguments
  */
 public static String [] getVMArgs() {
+	return getVMArgsHelper(true);
+}
+
+/**
+ * Returns command line arguments passed to the VM. Internally these are broken into
+ * optionString and extraInfo. This only returns the optionString part.
+ * <p>
+ * @return		a String array containing the optionString part of USER command line arguments
+ */
+public static String [] getAllVMArgs() {
+	return getVMArgsHelper(false);
+}
+
+/**
+ * Helper function for getVMArgs
+ * @param getUserArgsOnly Tells function if we wangt to return user arguments only
+ * @return		a String array containing the optionString part of command line arguments
+ */
+private static String [] getVMArgsHelper(boolean getUserArgsOnly) {
 	/*[PR CMVC 112580] Java cannot handle non-ascii character arguments */
-	if (cachedVMArgs == null) {
-		byte[][] byteArgs = getVMArgsImpl();
-		if (byteArgs == null) return null;
-		String[] result = new String[byteArgs.length];
-		for (int i=0; i<byteArgs.length; i++) {
-			if (byteArgs[i] == null) {
-				result[i] = null;
-			} else {
-				result[i] = Util.toString(byteArgs[i]);
-			}
+	byte[][] byteArgs = (getUserArgsOnly) ? getVMArgsImpl() : getAllVMArgsImpl();
+	if (byteArgs == null) return null;
+	String[] result = new String[byteArgs.length];
+	for (int i=0; i<byteArgs.length; i++) {
+		if (byteArgs[i] == null) {
+			result[i] = null;
+		} else {
+			result[i] = Util.toString(byteArgs[i]);
 		}
-		cachedVMArgs = result;
```

Comment:
Strange, I thought I left a comment...
The reason why we don't cache the VM arguments is because if we were to call getVMArgs() and getAllVMArgs() in the same java file, the function that is called second would return the results from the first.

Eg, if i called getVMArgs(), then getAllVMArgs() right after, getVMArgs() would return the user arguments only (since only the user arguments are cached)

Reply:
You could cache both sets: user args and all args.  If you want to be clever, you could store the disjoint sets (user and non-user) and return  their unions if you request all args.